### PR TITLE
Implement RPC-URI-V1-12 linter rule

### PR DESCRIPTION
**Pull request by AkhilaIlla**
This rule checks if tracked extensions resources exists or not and flags if exists

### CODE REVIEW

File: common/changes/@microsoft.azure/openapi-validator-rulesets/akhilailla-implement_rpc_uri_v1_v12_2023-09-13-23-08.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft.azure/openapi-validator-rulesets",
+      "comment": "implement rpc-uri-v1-12",
```
Comment (rkmanda) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/akhilailla-implement_rpc_uri_v1_v12_2023-09-13-23-08.json:5]: >implement rpc-uri-v1-12 [](http://example.com/codeflow?start=18&length=23)

change log needs to be user facing. pl take a look at the guidelines posted by Brett earlier

File: docs/tracked-extension-resources-are-not-allowed.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+{0} is an extension resource and the response schema in {1} operation includes location property. Extension resources of type tracked are not allowed.
+
+## Description
+
+Extension resources cannot be of type tracked.
+
+## How to fix the violation
+
+Remove "location" property from the property bag of an extension resource.
```
Comment (rkmanda) [inline docs/tracked-extension-resources-are-not-allowed.md:25]: >property bag [](http://example.com/codeflow?start=36&length=12)

location is a top level property and not in the property bag. change it to

Remove the top level "location" property from the model of the extension resource.

File: docs/tracked-extension-resources-are-not-allowed.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+- RPC-Uri-V1-12
+
+## Output Message
+
+{0} is an extension resource and the response schema in {1} operation includes location property. Extension resources of type tracked are not allowed.
+
+## Description
+
+Extension resources cannot be of type tracked.
+
```
Comment (rkmanda) [inline docs/tracked-extension-resources-are-not-allowed.md:22]: Extension resources are always considered to be proxy and cannot be of the tracked type.

File: docs/tracked-extension-resources-are-not-allowed.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ARM OpenAPI (swagger) specs
+
+## Related ARM Guideline Code
+
+- RPC-Uri-V1-12
+
+## Output Message
+
+{0} is an extension resource and the response schema in {1} operation includes location property. Extension resources of type tracked are not allowed.
+
```
Comment (rkmanda) [inline docs/tracked-extension-resources-are-not-allowed.md:18]: Can u add an item to standardize the format of all the md files? Some of them seem to have the output message while others may not be having this.

File: packages/rulesets/generated/spectral/az-arm.js [hunk 5]
```diff
@@ -182,6 +182,11 @@ const LATEST_VERSION_BY_COMMON_TYPES_FILENAME = new Map([
 function isLatestCommonTypesVersionForFile(version, fileName) {
     return LATEST_VERSION_BY_COMMON_TYPES_FILENAME.get(fileName) === version.toLowerCase();
 }
+const ExtensionResourceFullyQualifiedPathReg = new RegExp(".+/providers/.+/providers/.+$", "gi");
```
Comment (rkmanda) [inline packages/rulesets/generated/spectral/az-arm.js:185]: new line

Comment (rkmanda) [reply packages/rulesets/generated/spectral/az-arm.js:185]: isnt there another function that also checks the same?

File: packages/rulesets/src/spectral/functions/utils.ts [hunk 6]
```diff
@@ -12,6 +12,13 @@ export function isLatestCommonTypesVersionForFile(version: string, fileName: str
   return LATEST_VERSION_BY_COMMON_TYPES_FILENAME.get(fileName) === version.toLowerCase()
 }
 
+const ExtensionResourceFullyQualifiedPathReg = new RegExp(".+/providers/.+/providers/.+$", "gi")
+const ExtensionResourceReg = new RegExp("^/{\\w+}/providers/.+$", "gi")
+
+export function isPathOfExtensionResource(path: string) {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/utils.ts:18]: >isPathOfExtensionResource [](http://example.com/codeflow?start=16&length=25)

can u doublecheck that there isnt another function that does the same thing?

File: packages/rulesets/src/spectral/test/tracked-extension-resources-are-not-allowed.test.ts [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+beforeAll(async () => {
+  linter = await linterForRule("trackedExtensionResourcesAreNotAllowed")
+  return linter
+})
+
+test("trackedExtensionResourcesAreNotAllowed should find errors for tracked extension resources", () => {
+  const myOpenApiDocument = {
+    swagger: "2.0",
+    paths: {
+      "/{resourceUri}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates": {
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/test/tracked-extension-resources-are-not-allowed.test.ts:15]: >/{resourceUri}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates [](http://example.com/codeflow?start=7&length=76)

Add tests for the following variants of extension resource api paths


Without the leading slash
{resourceUri}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates

Explicit extension 1
/subscriptions/{subid}/resourceGroups/{rgName}/providers/Microsoft.Foo/resourceType/{resourceInstance}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates


Explicit extension 2
/subscriptions/{subid}/providers/Microsoft.Foo/resourceType/{resourceInstance}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates


Explicit extension 3
/subscriptions/{subid}/providers/Microsoft.Foo/resourceType/{resourceInstance}/providers/Microsoft.Bar/resourceType/{resourceInstance}/providers/Microsoft.EdgeProvisioning/deviceProvisioningStates

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: docs/tracked-extension-resources-are-not-allowed.md (hunk 4 above)
Comment (AkhilaIlla) [reply docs/tracked-extension-resources-are-not-allowed.md:18]: [https://msazure.visualstudio.com/One/_workitems/edit/25183186](https://msazure.visualstudio.com/One/_workitems/edit/25183186)

File: packages/rulesets/src/spectral/functions/utils.ts (hunk 6 above)
Comment (AkhilaIlla) [reply packages/rulesets/src/spectral/functions/utils.ts:18]: We have it defined in the legacy utils file, which I dont think can be used

File: common/changes/@microsoft.azure/openapi-validator-rulesets/akhilailla-implement_rpc_uri_v1_v12_2023-09-13-23-08.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft.azure/openapi-validator-rulesets",
+      "comment": "Added trackedExtensionResourcesAreNotAllowed rule that flags tracked extension resources, as extension resources may only be proxy",
```
Comment (rkmanda) [inline common/changes/@microsoft.azure/openapi-validator-rulesets/akhilailla-implement_rpc_uri_v1_v12_2023-09-13-23-08.json:5]: >Added [](http://example.com/codeflow?start=18&length=5)

Present tense

[https://rushjs.io/pages/best_practices/change_logs/#example-change-log-messages](https://rushjs.io/pages/best_practices/change_logs/#example-change-log-messages)

**Review by rkmanda — APPROVED** [review]
:shipit:

**Pull request merged**